# Analysis and visualization of crimes in Chicago

![](https://sites.psu.edu/aspsy/files/2017/03/chicago-u8nahk.jpg)

#### Authors: Karl Arnar Ægisson and Thomas Ari Bech

### Table of contents

1. Introduction
2. Required libraries
3. The data

### Introduction

/// https://en.wikipedia.org/wiki/Crime_in_Chicago

Crime in Chicago has been making headlines lately but in 2016, Chicago was responsible for nearly half of 2016's increase in homicides in the US, even though the nation's crime rates remain near historic lows.

What are the reasons for the high numbers in Chicago? Will the crime rates be likely to increase or decrease in the future?

In this notebook we will be trying to answer those questions.

### Required libraries

Along with Python we will be using several Python packages, these packages are:
* **NumPy** - Adds support for fast numerical array structure, along with helper functions.
* **pandas** - Provides a high-performance DataFrame structure.
* **matplotlib** - Plotting library for Python and NumPy.
* **Seaborn** - Advanced statistical plotting library.
* **scikit-learn** - Machine Learning package.

Let's import these packages:

In [55]:
# libraries we will need
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

#### The data

We will be using a data that is released by the City of Chicago. The data contains reported incidents of crime that occured in Chicago between 2013 to 2017. The data has the following attributes:


| **Attribute**   | **Description** |
| :------------  | :------------- |
| **ID** | Unique identifier for the record |
| **Case Number** | The Chicago Police Department RD Number (Records Division Number), which is unique to the incident |
| **Date** | Date when the incident occurred. this is sometimes a best estimate |
| **Block** | The partially redacted address where the incident occurred, placing it on the same block as the actual address |
| **IUCR** | The Illinois Unifrom Crime Reporting code. This is directly linked to the Primary Type and Description |
| **Primary Type** | The primary description of the IUCR code |
| **Description** | The secondary description of the IUCR code, a subcategory of the primary description |
| **Location Description** | Description of the location where the incident occurred |
| **Arrest** | Indicates whether an arrest was made |
| **Domestic** | Indicates whether the incident was domestic-related as defined by the Illinois Domestic Violence Act |
| **Beat** | Indicates the beat where the incident occurred. A beat is the smallest police geographic area – each beat has a dedicated police beat car |
| **District** | Indicates the police district where the incident occurred |
| **Ward** | The ward (City Council district) where the incident occurred |
| **Community Area** | Indicates the community area where the incident occurred. Chicago has 77 community areas |
| **FBI Code** | Indicates the crime classification as outlined in the FBI's National Incident-Based Reporting System (NIBRS) |
| **X Coordinate** | The x coordinate of the location where the incident occurred in State Plane Illinois East NAD 1983 projection |
| **Y Coordinate** | The y coordinate of the location where the incident occurred in State Plane Illinois East NAD 1983 projection |
| **Year** | Year the incident occurred |
| **Updated On** | Date and time the record was last updated |
| **Latitude** | The latitude of the location where the incident occurred |
| **Longitude** | The longitude of the location where the incident occurred |
| **Location** | The location where the incident occurred in a format that allows for creation of maps and other geographic operations on this data portal |


Since the data contains several attributes that we will not be using, we will drop these attributes from the DataFrame:
* Case Number
* IUCR
* Description
* Beat
* District
* Ward
* FBI Code
* X Coordinate
* Y Coordinate
* Year
* Updated On

Which will leave us with these attributes:
* ID
* Date
* Block
* Primary Type
* Location Description
* Arrest
* Domestic
* Community Area
* Latitude
* Longitude
* Location

In [56]:
df = pd.read_csv('Chicago_Crimes_2012_to_2017.csv', na_values=['NA'])
df.info()
df.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1456714 entries, 0 to 1456713
Data columns (total 23 columns):
Unnamed: 0              1456714 non-null int64
ID                      1456714 non-null int64
Case Number             1456713 non-null object
Date                    1456714 non-null object
Block                   1456714 non-null object
IUCR                    1456714 non-null object
Primary Type            1456714 non-null object
Description             1456714 non-null object
Location Description    1455056 non-null object
Arrest                  1456714 non-null bool
Domestic                1456714 non-null bool
Beat                    1456714 non-null int64
District                1456713 non-null float64
Ward                    1456700 non-null float64
Community Area          1456674 non-null float64
FBI Code                1456714 non-null object
X Coordinate            1419631 non-null float64
Y Coordinate            1419631 non-null float64
Year                    1

Unnamed: 0        ID Case Number                    Date                Block  IUCR            Primary Type              Description Location Description  Arrest              ...                Ward  Community Area  FBI Code  X Coordinate  Y Coordinate  Year              Updated On   Latitude  Longitude                       Location
0           3  10508693    HZ250496  05/03/2016 11:40:00 PM   013XX S SAWYER AVE  0486                 BATTERY  DOMESTIC BATTERY SIMPLE            APARTMENT    True              ...                24.0            29.0       08B     1154907.0     1893681.0  2016  05/10/2016 03:56:50 PM  41.864073 -87.706819  (41.864073157, -87.706818608)
1          89  10508695    HZ250409  05/03/2016 09:40:00 PM   061XX S DREXEL AVE  0486                 BATTERY  DOMESTIC BATTERY SIMPLE            RESIDENCE   False              ...                20.0            42.0       08B     1183066.0     1864330.0  2016  05/10/2016 03:56:50 PM  41.782922 -87.604363   (41.782921527, -87.60436317)
2         197  10508697    HZ250503  05/03/2016 11:31:00 PM  053XX W CHICAGO AVE  0470  PUBLIC PEACE VIOLATION         RECKLESS CONDUCT               STREET   False              ...                37.0            25.0        24     1140789.0     1904819.0  2016  05/10/2016 03:56:50 PM  41.894908 -87.758372  (41.894908283, -87.758371958)
3         673  10508698    HZ250424  05/03/2016 10:10:00 PM    049XX W FULTON ST  0460                 BATTERY                   SIMPLE             SIDEWALK   False              ...                28.0            25.0       08B     1143223.0     1901475.0  2016  05/10/2016 03:56:50 PM  41.885687 -87.749516  (41.885686845, -87.749515983)
4         911  10508699    HZ250455  05/03/2016 10:00:00 PM    003XX N LOTUS AVE  0820                   THEFT           $500 AND UNDER            RESIDENCE   False              ...                28.0            25.0        06     1139890.0     1901675.0  2016  05/10/2016 03:56:50 PM  41.886297 -87.761751  (41.886297242, -87.761750709)

[5 rows x 23 columns]

We can see that there are a total of 1.456.714 records in the data.

Let's drop those features that we aren't going to use and let's see the DataFrame again.

In [57]:
df.drop(['Unnamed: 0', 'Case Number', 'IUCR', 'Description', 'Beat', 'District', 'Ward', 'FBI Code', 'X Coordinate', 'Y Coordinate', 'Year', 'Updated On'], inplace=True, axis=1)
df.info()
df.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1456714 entries, 0 to 1456713
Data columns (total 11 columns):
ID                      1456714 non-null int64
Date                    1456714 non-null object
Block                   1456714 non-null object
Primary Type            1456714 non-null object
Location Description    1455056 non-null object
Arrest                  1456714 non-null bool
Domestic                1456714 non-null bool
Community Area          1456674 non-null float64
Latitude                1419631 non-null float64
Longitude               1419631 non-null float64
Location                1419631 non-null object
dtypes: bool(2), float64(3), int64(1), object(5)
memory usage: 102.8+ MB


ID                    Date                Block            Primary Type Location Description  Arrest  Domestic  Community Area   Latitude  Longitude                       Location
0  10508693  05/03/2016 11:40:00 PM   013XX S SAWYER AVE                 BATTERY            APARTMENT    True      True            29.0  41.864073 -87.706819  (41.864073157, -87.706818608)
1  10508695  05/03/2016 09:40:00 PM   061XX S DREXEL AVE                 BATTERY            RESIDENCE   False      True            42.0  41.782922 -87.604363   (41.782921527, -87.60436317)
2  10508697  05/03/2016 11:31:00 PM  053XX W CHICAGO AVE  PUBLIC PEACE VIOLATION               STREET   False     False            25.0  41.894908 -87.758372  (41.894908283, -87.758371958)
3  10508698  05/03/2016 10:10:00 PM    049XX W FULTON ST                 BATTERY             SIDEWALK   False     False            25.0  41.885687 -87.749516  (41.885686845, -87.749515983)
4  10508699  05/03/2016 10:00:00 PM    003XX N LOTUS AVE                   THEFT            RESIDENCE   False      True            25.0  41.886297 -87.761751  (41.886297242, -87.761750709)

That leaves us with a total of 11 columns (features).

Next step is to check if the data contains any duplicates, and if they do, drop those rows. Since we know that the ID is a unique identifier for the record, we will check if there are duplicate ID's.

In [58]:
bool_s = df.duplicated(keep=False)
df[bool_s].sort_values("ID")

Empty DataFrame
Columns: [ID, Date, Block, Primary Type, Location Description, Arrest, Domestic, Community Area, Latitude, Longitude, Location]
Index: []

Great! We can see that the data contains no duplicate records according to the ID.

Let's check how many non N/A there is for every column:

In [61]:
df.count()

ID                      1456714
Date                    1456714
Block                   1456714
Primary Type            1456714
Location Description    1455056
Arrest                  1456714
Domestic                1456714
Community Area          1456674
Latitude                1419631
Longitude               1419631
Location                1419631
dtype: int64

Here we can see that there are 5 columns that contain less non N/A values that the other six. Let's work on either removing those rows that have N/A according to the 5 columns, or trying to check if we can fix the N/A value.
  
We can see that Latitude, Longitude and Location columns all have the same non N/A values, we can drop those N/A values since we will be needing to use those attributes. That would trim our data down to 1.419.631 rows with 7 columns containing no N/A values.

In [62]:
df = df.dropna(subset=['Latitude', 'Longitude', 'Location'])
df.isnull().sum()

ID                         0
Date                       0
Block                      0
Primary Type               0
Location Description    1226
Arrest                     0
Domestic                   0
Community Area            25
Latitude                   0
Longitude                  0
Location                   0
dtype: int64

Next is "community area". We can see that it contains 25 NA values. We can view those rows that contain NA value in "community area" by creating a subset and displaying the null values in the subset.

In [63]:
df_subset = df[['Community Area']]
dfCom = df[df_subset.isnull().any(axis=1)]
dfCom

ID                    Date                    Block Primary Type            Location Description  Arrest  Domestic  Community Area   Latitude  Longitude                       Location
77621   8421956  01/01/2012 11:15:00 AM       083XX S HOLLAND RD        THEFT                DEPARTMENT STORE    True     False             NaN  41.742991 -87.632026  (41.742991296, -87.632025815)
80166   8427468  01/05/2012 12:30:00 PM         013XX W 105TH ST        THEFT                           OTHER   False     False             NaN  41.703127 -87.656698  (41.703127069, -87.656697951)
84948   8436090  01/11/2012 07:00:00 AM         013XX W 105TH ST        THEFT                           OTHER   False     False             NaN  41.703127 -87.656698  (41.703127069, -87.656697951)
87270   8440227  01/14/2012 06:51:00 PM       053XX S HARLEM AVE    NARCOTICS                          STREET    True     False             NaN  41.795354 -87.801608  (41.795354198, -87.801608429)
91421   8447192  01/17/2012 07:45:00 AM        032XX W MONROE ST        THEFT                          STREET   False     False             NaN  41.879634 -87.707974  (41.879633871, -87.707973888)
100161  8461905  01/30/2012 04:30:00 PM        052XX W OLIVE AVE     BURGLARY               CONSTRUCTION SITE   False     False             NaN  41.983996 -87.759958  (41.983996419, -87.759958171)
102034  8465236  01/04/2012 06:03:00 PM  099XX S CONFIDENTIAL ST    NARCOTICS          VEHICLE NON-COMMERCIAL   False     False             NaN  36.619446 -91.686566  (36.619446395, -91.686565684)
104058  8468856  02/04/2012 01:40:00 PM       084XX S HOLLAND RD      BATTERY  PARKING LOT/GARAGE(NON.RESID.)   False      True             NaN  41.740990 -87.631971   (41.740990358, -87.63197064)
104816  8469806  02/02/2012 10:30:00 AM      057XX N LATROBE AVE      BATTERY                       RESIDENCE   False      True             NaN  41.985080 -87.760335  (41.985079514, -87.760334655)
110635  8479349  02/13/2012 12:10:00 AM         055XX W HOMER ST      BATTERY             MOVIE HOUSE/THEATER   False     False             NaN  41.915055 -87.764231  (41.915055487, -87.764231386)
...         ...                     ...                      ...          ...                             ...     ...       ...             ...        ...        ...                            ...
114413  8485577  02/17/2012 01:10:00 PM        032XX W MONROE ST      ASSAULT         SCHOOL, PUBLIC, GROUNDS    True     False             NaN  41.879647 -87.706365  (41.879647055, -87.706365394)
115345  8486711  02/18/2012 03:25:00 PM       083XX S HOLLAND RD        THEFT              GROCERY FOOD STORE   False     False             NaN  41.742991 -87.632026  (41.742991296, -87.632025815)
115589  8487005  02/18/2012 10:00:00 PM        032XX W MONROE ST      ROBBERY                        SIDEWALK   False     False             NaN  41.879647 -87.706769  (41.879646516, -87.706769327)
117388  8489544  02/21/2012 12:15:00 PM        032XX W MONROE ST      ASSAULT        SCHOOL, PUBLIC, BUILDING    True     False             NaN  41.879634 -87.707974  (41.879633871, -87.707973888)
118479  8491415  02/03/2012 09:00:00 AM       084XX S HOLLAND RD        THEFT              SMALL RETAIL STORE    True     False             NaN  41.740919 -87.631968  (41.740918989, -87.631967753)
120969  8495544  02/25/2012 02:00:00 PM         055XX W HOMER ST        THEFT          VEHICLE NON-COMMERCIAL   False     False             NaN  41.915055 -87.764180   (41.915055232, -87.76417995)
121757  8496554  02/26/2012 06:35:00 PM       083XX S HOLLAND RD      ASSAULT  PARKING LOT/GARAGE(NON.RESID.)   False     False             NaN  41.742991 -87.632026  (41.742991296, -87.632025815)
122919  8498547  02/27/2012 01:32:00 PM       085XX S HOLLAND RD      ASSAULT              SMALL RETAIL STORE   False     False             NaN  41.739824 -87.631939  (41.739823836, -87.631939365)
174204  8582892  04/09/2012 10:05:00 AM  099XX S CONFIDENTIAL ST    NARCO

We can see that the 25 NA values all contain valid information except for the community area. We could drop those rows but let's try to maybe fill in the data.

We can see that the 'Block' column contains an address, we can use that address to check if there is another row that contains the same address with the 'Community Area' column filled.

Let's take three random addresses and check the rows that contain that address in the 'block' column.

In [64]:
df.loc[df['Block'] == "083XX S HOLLAND RD"]

ID                    Date               Block        Primary Type            Location Description  Arrest  Domestic  Community Area   Latitude  Longitude                       Location
2204     10515160  05/06/2016 09:00:00 AM  083XX S HOLLAND RD  DECEPTIVE PRACTICE                            BANK   False     False            44.0  41.742991 -87.632026  (41.742991296, -87.632025815)
38855     9931616  01/19/2015 08:44:00 AM  083XX S HOLLAND RD               THEFT              VEHICLE-COMMERCIAL   False     False            44.0  41.742656 -87.632018  (41.742656445, -87.632018466)
65889     9983362  03/04/2015 04:15:00 PM  083XX S HOLLAND RD               THEFT  PARKING LOT/GARAGE(NON.RESID.)   False     False            44.0  41.742656 -87.632018  (41.742656445, -87.632018466)
77621     8421956  01/01/2012 11:15:00 AM  083XX S HOLLAND RD               THEFT                DEPARTMENT STORE    True     False             NaN  41.742991 -87.632026  (41.742991296, -87.632025815)
115345    8486711  02/18/2012 03:25:00 PM  083XX S HOLLAND RD               THEFT              GROCERY FOOD STORE   False     False             NaN  41.742991 -87.632026  (41.742991296, -87.632025815)
121757    8496554  02/26/2012 06:35:00 PM  083XX S HOLLAND RD             ASSAULT  PARKING LOT/GARAGE(NON.RESID.)   False     False             NaN  41.742991 -87.632026  (41.742991296, -87.632025815)
156690    8554222  04/05/2012 08:16:00 PM  083XX S HOLLAND RD               THEFT               CONVENIENCE STORE    True     False            44.0  41.742991 -87.632026  (41.742991296, -87.632025815)
189072    8607408  05/10/2012 09:50:00 PM  083XX S HOLLAND RD             ROBBERY                        SIDEWALK    True     False            44.0  41.742947 -87.632023  (41.742947368, -87.632022629)
190639    8609766  05/12/2012 04:55:00 PM  083XX S HOLLAND RD             ROBBERY              SMALL RETAIL STORE   False     False            44.0  41.742991 -87.632026  (41.742991296, -87.632025815)
192806    8613158  05/11/2012 04:00:00 PM  083XX S HOLLAND RD               THEFT                           OTHER   False     False            44.0  41.742991 -87.632026  (41.742991296, -87.632025815)
...           ...                     ...                 ...                 ...                             ...     ...       ...             ...        ...        ...                            ...
1353623  10666140  09/01/2016 09:30:00 PM  083XX S HOLLAND RD               THEFT  PARKING LOT/GARAGE(NON.RESID.)   False     False            44.0  41.742991 -87.632026  (41.742991296, -87.632025815)
1370713  10691233  09/22/2016 07:00:00 PM  083XX S HOLLAND RD               THEFT  PARKING LOT/GARAGE(NON.RESID.)   False     False            44.0  41.742991 -87.632026  (41.742991296, -87.632025815)
1395762  10730818  10/24/2016 02:00:00 PM  083XX S HOLLAND RD               THEFT                   ATHLETIC CLUB   False     False            44.0  41.742991 -87.632026  (41.742991296, -87.632025815)
1409099  10749475  11/11/2016 09:30:00 AM  083XX S HOLLAND RD               THEFT                   ATHLETIC CLUB   False     False            44.0  41.742656 -87.632018  (41.742656445, -87.632018466)
1412196  10754125  11/15/2016 02:30:00 PM  083XX S HOLLAND RD               THEFT           MEDICAL/DENTAL OFFICE   False     False            44.0  41.742991 -87.632026  (41.742991296, -87.632025815)
1412739  10754804  11/16/2016 03:00:00 AM  083XX S HOLLAND RD             BATTERY                   ATHLETIC CLUB   False     False            44.0  41.742656 -87.632018  (41.742656445, -87.632018466)
1415106  10758324  11/19/2016 07:00:00 AM  083XX S HOLLAND RD               THEFT                   ATHLETIC CLUB   False     False            44.0  41.742991 -87.632026  (41.742991296, -87.632025815)
1419880  10764941  11/26/2016 02:58:00 PM  083XX S HOLLAND RD             BATTERY                           OTHER   False     False            44.0  41.742991 -87.632026  (41.742991296, -87.632025

In [65]:
df.loc[df['Block'] == "055XX W HOMER ST"]

ID                    Date             Block       Primary Type            Location Description  Arrest  Domestic  Community Area   Latitude  Longitude                       Location
29698     9913376  01/02/2015 08:00:00 PM  055XX W HOMER ST            BATTERY                           OTHER   False     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
32901     9920703  01/09/2015 11:00:00 PM  055XX W HOMER ST              THEFT                           OTHER   False     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
52906     9958748  02/10/2015 02:00:00 PM  055XX W HOMER ST              THEFT             MOVIE HOUSE/THEATER   False     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
55318     9962847  02/14/2015 10:02:00 PM  055XX W HOMER ST  CRIMINAL TRESPASS             MOVIE HOUSE/THEATER    True     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
64096     9979313  03/01/2015 06:45:00 PM  055XX W HOMER ST              THEFT             MOVIE HOUSE/THEATER   False     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
110635    8479349  02/13/2012 12:10:00 AM  055XX W HOMER ST            BATTERY             MOVIE HOUSE/THEATER   False     False             NaN  41.915055 -87.764231  (41.915055487, -87.764231386)
112950    8483029  02/15/2012 07:20:00 PM  055XX W HOMER ST    CRIMINAL DAMAGE             MOVIE HOUSE/THEATER   False     False             NaN  41.915055 -87.764180   (41.915055232, -87.76417995)
120969    8495544  02/25/2012 02:00:00 PM  055XX W HOMER ST              THEFT          VEHICLE NON-COMMERCIAL   False     False             NaN  41.915055 -87.764180   (41.915055232, -87.76417995)
183850    8598943  05/05/2012 04:00:00 PM  055XX W HOMER ST    CRIMINAL DAMAGE          VEHICLE NON-COMMERCIAL   False     False            25.0  41.915053 -87.764257   (41.91505287, -87.764257128)
244162    8692286  07/05/2012 01:45:00 PM  055XX W HOMER ST              THEFT             MOVIE HOUSE/THEATER   False     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
...           ...                     ...               ...                ...                             ...     ...       ...             ...        ...        ...                            ...
1380668  10706238  10/05/2016 06:30:00 PM  055XX W HOMER ST              THEFT  PARKING LOT/GARAGE(NON.RESID.)   False     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
1380737  10706335  10/05/2016 06:30:00 PM  055XX W HOMER ST              THEFT  PARKING LOT/GARAGE(NON.RESID.)   False     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
1380760  10706375  10/05/2016 06:30:00 PM  055XX W HOMER ST              THEFT  PARKING LOT/GARAGE(NON.RESID.)   False     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
1383817  10710548  10/09/2016 08:10:00 PM  055XX W HOMER ST    CRIMINAL DAMAGE             MOVIE HOUSE/THEATER   False     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
1395680  10729204  10/24/2016 09:40:00 PM  055XX W HOMER ST              THEFT                           OTHER   False     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
1407429  10748821  11/08/2016 08:30:00 PM  055XX W HOMER ST              THEFT             MOVIE HOUSE/THEATER   False     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
1409265  10749829  11/11/2016 06:40:00 PM  055XX W HOMER ST    CRIMINAL DAMAGE  PARKING LOT/GARAGE(NON.RESID.)   False     False            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
1409315  10749943  11/11/2016 04:45:00 PM  055XX W HOMER ST            BATTERY             MOVIE HOUSE/THEATER   False      True            25.0  41.915055 -87.764180   (41.915055232, -87.76417995)
1417121  10761157  11/22/2016 01:00:00 AM  055XX W HOME

In [66]:
df.loc[df['Block'] == "099XX S CONFIDENTIAL"]

ID                    Date                 Block Primary Type Location Description  Arrest  Domestic  Community Area   Latitude  Longitude                       Location
188517  8606659  04/13/2012 11:00:00 AM  099XX S CONFIDENTIAL    NARCOTICS             SIDEWALK   False     False             NaN  36.619446 -91.686566  (36.619446395, -91.686565684)

We can see that "083XX S HOLLAND RD" contains 112 rows and most of them have "44" in the 'Community Area' column.

Same story with "055XX W HOMER ST", most of the 78 rows have "25" in the 'Community Area' column.

However "099XX S CONFIDENTIAL", only contains the NA value.

Based on this observation, we can iterate through the rows in the dataframe we created prior, called dfCom and contains the subset of our main dataframe (df) that contains all the rows that have NA value in the community area column.

Then we can find the most occurence of a value in community area for that block and replace the NA value with the mode value.

In [67]:
for index, row in dfCom.iterrows():
    df.at[index, 'Community Area'] = df.loc[df['Block'] == row['Block']].mode()["Community Area"][:1]

df_subset = df[['Community Area']]
dfCom = df[df_subset.isnull().any(axis=1)]
dfCom

ID                    Date                    Block Primary Type    Location Description  Arrest  Domestic  Community Area   Latitude  Longitude                       Location
102034  8465236  01/04/2012 06:03:00 PM  099XX S CONFIDENTIAL ST    NARCOTICS  VEHICLE NON-COMMERCIAL   False     False             NaN  36.619446 -91.686566  (36.619446395, -91.686565684)
104816  8469806  02/02/2012 10:30:00 AM      057XX N LATROBE AVE      BATTERY               RESIDENCE   False      True             NaN  41.985080 -87.760335  (41.985079514, -87.760334655)
174204  8582892  04/09/2012 10:05:00 AM  099XX S CONFIDENTIAL ST    NARCOTICS               RESIDENCE   False     False             NaN  36.619446 -91.686566  (36.619446395, -91.686565684)
188517  8606659  04/13/2012 11:00:00 AM     099XX S CONFIDENTIAL    NARCOTICS                SIDEWALK   False     False             NaN  36.619446 -91.686566  (36.619446395, -91.686565684)

We can see that our approach worked on all but 4 records. Looking at the data, we can safely assume that those rows that contain CONFIDENTIAL in them are values that are CONFIDENTIAL, this is confirmed if we lookup the location value which shows the location at some remote part outside of Illinois. We will drop these rows since they don't contain a valid location or block information.

That leaves us with one NA value which has "057XX N LATROBE AVE". We can drop the row or we can try to find out a way to fill in the community area.

Let's try to find if there are other blocks that contain "XX N LATROBE AVE" in the 'Block' column.

We can do that by using str.contains, which works much like 'LIKE' in SQL.

In [68]:
pd.set_option('display.max_rows', 1500) # changing so we can see all values
df[df['Block'].str.contains('XX N LATROBE AVE')].sort_values("Block")

ID                    Date                Block                      Primary Type            Location Description  Arrest  Domestic  Community Area   Latitude  Longitude                       Location
757278    9527600  03/11/2014 07:00:00 PM  001XX N LATROBE AVE                       SEX OFFENSE                       APARTMENT   False     False            25.0  41.883301 -87.756394  (41.883301178, -87.756393849)
405403    8948456  09/01/2012 08:00:00 AM  001XX N LATROBE AVE                             THEFT                          STREET   False     False            25.0  41.883153 -87.756388  (41.883152958, -87.756387838)
915448    9797385  09/28/2014 05:20:00 AM  001XX N LATROBE AVE                             THEFT                       RESIDENCE   False      True            25.0  41.883024 -87.756382  (41.883023947, -87.756381654)
979450    9906867  12/28/2014 03:35:00 PM  001XX N LATROBE AVE                          BURGLARY                       RESIDENCE    True     False            25.0  41.883534 -87.756403  (41.883534484, -87.756402767)
340637    8844762  10/14/2012 03:10:00 PM  001XX N LATROBE AVE                   CRIMINAL DAMAGE                          STREET   False      True            25.0  41.883449 -87.756400   (41.883449397, -87.75639986)
406306    8949541  12/30/2012 02:11:00 PM  001XX N LATROBE AVE                     OTHER OFFENSE                          STREET   False     False            25.0  41.883301 -87.756394  (41.883301178, -87.756393849)
1113433  10225328  09/06/2015 03:30:00 AM  001XX N LATROBE AVE                           BATTERY                           ALLEY   False     False            25.0  41.882961 -87.756382  (41.882960832, -87.756382222)
569805    9213934  07/13/2013 04:20:00 AM  001XX N LATROBE AVE            PUBLIC PEACE VIOLATION                        SIDEWALK    True     False            25.0  41.883364 -87.756397  (41.883364311, -87.756396953)
572015    9217098  07/14/2013 06:00:00 AM  001XX N LATROBE AVE                           BATTERY         RESIDENCE PORCH/HALLWAY   False      True            25.0  41.883175 -87.756388   (41.883174911, -87.75638764)
1303784  10582497  07/01/2016 02:40:00 PM  001XX N LATROBE AVE                           BATTERY                        SIDEWALK   False      True            25.0  41.883534 -87.756403  (41.883534484, -87.756402767)
113018    8483113  02/15/2012 11:36:00 PM  001XX N LATROBE AVE                           BATTERY                       RESIDENCE   False      True            25.0  41.883153 -87.756388  (41.883152958, -87.756387838)
414612    8962991  01/10/2013 02:40:00 AM  001XX N LATROBE AVE                           BATTERY                       APARTMENT   False      True            25.0  41.883430 -87.756400  (41.883430189, -87.756400033)
111203    8480116  02/13/2012 04:45:00 PM  001XX N LATROBE AVE                           BATTERY                        SIDEWALK   False     False            25.0  41.883534 -87.756403  (41.883534484, -87.756402767)
111152    8480054  02/13/2012 04:08:00 PM  001XX N LATROBE AVE                           ASSAULT                       RESIDENCE    True     False            25.0  41.883534 -87.756403  (41.883534484, -87.756402767)
108914    8477163  02/10/2012 02:30:00 PM  001XX N LATROBE AVE                         NARCOTICS   RESIDENTIAL YARD (FRONT/BACK)    True     False            25.0  41.883449 -87.756400   (41.883449397, -87.75639986)
108593    8476469  01/17/2012 09:00:00 AM  001XX N LATROBE AVE                             THEFT                       RESIDENCE   False     False            25.0  41.883430 -87.756400  (41.883430189, -87.756400033)
225667    8663765  06/17/2012 12:15:00 PM  001XX N LATROBE AVE                           BATTERY                          STREET    True      True            25.0  41.883109 -87.756388  (41.883109052, -87.756388233)
1116160  10229842  09/09/2015 06:27:00 PM  001XX N LATROBE AVE                         NARCOTICS                        SIDEWAL

We can see that there are 1.316 rows that contain "XX N LATROBE AVE" in the name! The rows contain three different values in the 'Community Area' column, along with our NA value.

Looking at the data we can see that there are trends in the data that we can look at to try to classify the NA value.

Blocks 001 to 018 contain "25" in the 'Community Area' column.
Blocks 019 to 023 contain "19" in the 'Community Area' column.
Blocks 052 to 056 contain "11" in the 'Community Area' column.

Based on these we must conclude that block 057 community area must be "11". We can verify this by looking up the location of 056 and 057 and checking if they are close to each other.

Let's change the Community Area to 11 for that row and then dropping the rows that contained the "(36.619446395, -91.686565684)" location which we found out was bogus.

In [70]:
pd.set_option('display.max_rows', 20) # changing back
df.at[104816, 'Community Area'] = 11.0
df = df.dropna(subset=['Community Area'])
df.isnull().sum()

ID                         0
Date                       0
Block                      0
Primary Type               0
Location Description    1226
Arrest                     0
Domestic                   0
Community Area             0
Latitude                   0
Longitude                  0
Location                   0
dtype: int64

This leaves us with the Location Description, we can see that it contains 1226 NA values, let's check the null values there

In [71]:
df_subset = df[['Location Description']]
df[df_subset.isnull().any(axis=1)]

ID                    Date                   Block        Primary Type Location Description  Arrest  Domestic  Community Area   Latitude  Longitude                       Location
98       10509988  05/03/2016 04:35:00 PM   029XX N SHEFFIELD AVE  DECEPTIVE PRACTICE                  NaN    True     False             6.0  41.934969 -87.653937  (41.934968944, -87.653936849)
133      10511522  05/03/2016 08:00:00 AM        033XX N Page Ave  DECEPTIVE PRACTICE                  NaN   False     False            17.0  41.940042 -87.827625  (41.940042028, -87.827625466)
471      10509989  05/04/2016 12:00:00 PM     047XX N KIMBALL AVE  DECEPTIVE PRACTICE                  NaN   False     False            14.0  41.967316 -87.713331   (41.96731629, -87.713331369)
690      10511523  05/04/2016 10:00:00 AM   035XX N LAKE SHORE DR  DECEPTIVE PRACTICE                  NaN   False     False             6.0  41.948068 -87.643126  (41.948068465, -87.643126381)
708      10512795  05/04/2016 11:30:00 AM    116XX S CARPENTER ST  DECEPTIVE PRACTICE                  NaN   False     False            53.0  41.682615 -87.648153  (41.682614691, -87.648152509)
909      10509982  04/29/2015 01:20:00 PM      026XX N KEDZIE AVE  DECEPTIVE PRACTICE                  NaN   False     False            22.0  41.928593 -87.707738   (41.92859251, -87.707738486)
1930     10512796  05/06/2016 09:45:00 AM       083XX S BAKER AVE  DECEPTIVE PRACTICE                  NaN   False     False            46.0  41.742589 -87.548984  (41.742588895, -87.548983923)
3698     10479906  04/08/2016 12:02:00 AM  016XX S MARSHFIELD AVE       OTHER OFFENSE                  NaN   False     False            31.0  41.859224 -87.666879  (41.859224265, -87.666878855)
3831     10486305  02/05/2015 02:15:00 AM     037XX W DICKENS AVE  DECEPTIVE PRACTICE                  NaN   False     False            22.0  41.919069 -87.720284  (41.919069102, -87.720283809)
4472     10518589  05/04/2016 02:00:00 PM         008XX E 79TH ST  DECEPTIVE PRACTICE                  NaN   False     False            69.0  41.751293 -87.604637  (41.751292524, -87.604636854)
...           ...                     ...                     ...                 ...                  ...     ...       ...             ...        ...        ...                            ...
1442020  10798921  12/29/2016 09:25:00 AM       058XX S DAMEN AVE  DECEPTIVE PRACTICE                  NaN   False     False            67.0  41.787289 -87.674115  (41.787289243, -87.674114635)
1443720  10803962  12/31/2016 12:00:00 PM       036XX N ODELL AVE  DECEPTIVE PRACTICE                  NaN   False     False            17.0  41.945842 -87.810678  (41.945841515, -87.810677977)
1444691  10806548  12/27/2016 06:15:00 PM         035XX W 77TH PL  DECEPTIVE PRACTICE                  NaN   False     False            70.0  41.752413 -87.710739  (41.752413176, -87.710738522)
1444730  10807785  12/29/2016 06:10:00 PM         015XX N WOOD ST  DECEPTIVE PRACTICE                  NaN   False     False            24.0  41.910111 -87.672562  (41.910110679, -87.672562212)
1446807  10811692  12/20/2016 03:05:00 PM     002XX N LA SALLE ST  DECEPTIVE PRACTICE                  NaN   False     False            32.0  41.886440 -87.632468  (41.886439939, -87.632467972)
1446808  10811712  12/24/2016 12:50:00 AM        0000X N STATE ST  DECEPTIVE PRACTICE                  NaN   False     False            32.0  41.882471 -87.627848  (41.882470918, -87.627847625)
1446809  10811718  12/27/2016 12:00:00 PM  051XX S HYDE PARK BLVD  DECEPTIVE PRACTICE                  NaN   False     False            41.0  41.801308 -87.584691  (41.801307936, -87.584690956)
1449361  10814232  12/22/2016 06:00:00 AM      044XX S Pulaski Rd  DECEPTIVE PRACTICE                  NaN   False     False            57.0  41.812611 -87.723765  (41.812610526, -87.723765071)
1449403  10816322  12/19/2016 06:00:00 AM        026XX N CLARK ST  DECEPTIVE PRACTICE                  NaN   False     False           

There is nothing that we can base the Location Description on so here we will just fill the NA values with 'OTHER', and let's check how many null values are left.

In [73]:
df = df.fillna('OTHER')
df.isnull().sum()

ID                      0
Date                    0
Block                   0
Primary Type            0
Location Description    0
Arrest                  0
Domestic                0
Community Area          0
Latitude                0
Longitude               0
Location                0
dtype: int64

Now we have dealt with all the NA values, let's check out how many rows we have left

In [75]:
df.count()

ID                      1419628
Date                    1419628
Block                   1419628
Primary Type            1419628
Location Description    1419628
Arrest                  1419628
Domestic                1419628
Community Area          1419628
Latitude                1419628
Longitude               1419628
Location                1419628
dtype: int64

We are left with 1.419.628 rows that contain no NA and that we can work with. Since we will be using date, let's convert it to a better format along with checking if it's between 01/01/2012 to 31/12/2017.


In [77]:
df['Date'] = pd.to_datetime(df['Date'], format='%m/%d/%Y %I:%M:%S %p')

df.sort_values('Date')

ID                Date                      Block                Primary Type Location Description  Arrest  Domestic  Community Area   Latitude  Longitude                       Location
382599   8911136 2012-01-01 00:00:00            040XX W 26TH ST          DECEPTIVE PRACTICE  TAVERN/LIQUOR STORE   False     False            30.0  41.844246 -87.725245  (41.844246118, -87.725244851)
130627   8511404 2012-01-01 00:00:00     064XX S SACRAMENTO AVE  OFFENSE INVOLVING CHILDREN            RESIDENCE    True      True            66.0  41.776618 -87.698204  (41.776617923, -87.698204428)
130646   8511440 2012-01-01 00:00:00            030XX W 64TH ST                       THEFT            RESIDENCE   False     False            66.0  41.777210 -87.699031   (41.777209768, -87.69903082)
77841    8422323 2012-01-01 00:00:00         029XX N ELSTON AVE                 SEX OFFENSE            RESIDENCE   False     False            21.0  41.935170 -87.692172  (41.935169613, -87.692172006)
490105   9087981 2012-01-01 00:00:00  008XX S INDEPENDENCE BLVD                       THEFT            RESIDENCE   False     False            27.0  41.870559 -87.719668  (41.870558601, -87.719668116)
589438   9245976 2012-01-01 00:00:00          056XX S MOZART ST         CRIM SEXUAL ASSAULT            RESIDENCE    True      True            63.0  41.791114 -87.694948  (41.791114438, -87.694948055)
77385    8421563 2012-01-01 00:00:00         033XX N HALSTED ST                       THEFT        BAR OR TAVERN   False     False             6.0  41.942351 -87.649345  (41.942351064, -87.649344776)
723953   9467940 2012-01-01 00:00:00         015XX S KEELER AVE                       THEFT            RESIDENCE   False     False            29.0  41.860113 -87.729881  (41.860113198, -87.729881467)
492396   9091480 2012-01-01 00:00:00  055XX S COTTAGE GROVE AVE                       THEFT            RESIDENCE   False     False            41.0  41.794253 -87.606211  (41.794253414, -87.606211387)
378271   8903224 2012-01-01 00:00:00           013XX W 112TH PL                       THEFT            RESIDENCE   False     False            75.0  41.689349 -87.656048  (41.689349439, -87.656047862)
...          ...                 ...                        ...                         ...                  ...     ...       ...             ...        ...        ...                            ...
1453793    23080 2017-01-14 19:34:00            055XX W RICE ST                    HOMICIDE               STREET   False     False            25.0  41.895772 -87.764223  (41.895771678, -87.764222641)
1454419    23081 2017-01-15 02:00:00            047XX W POLK ST                    HOMICIDE                ALLEY   False     False            25.0  41.870473 -87.743622  (41.870472791, -87.743622093)
1454420    23082 2017-01-15 05:50:00       026XX W VAN BUREN ST                    HOMICIDE               STREET   False     False            28.0  41.876028 -87.692069  (41.876028038, -87.692069498)
1455109    23084 2017-01-16 02:05:00        008XX N OAKLEY BLVD                    HOMICIDE               STREET   False     False            24.0  41.896015 -87.684402  (41.896015204, -87.684402372)
1455111    23086 2017-01-16 12:11:00         048XX W HUBBARD ST                    HOMICIDE                PORCH   False     False            25.0  41.888677 -87.746698  (41.888677432, -87.746697587)
1455110    23085 2017-01-16 13:35:00        004XX S CENTRAL AVE                    HOMICIDE               STREET   False     False            25.0  41.874393 -87.764647   (41.874393341, -87.76464733)
1455885    23088 2017-01-17 23:42:00        001XX N CENTRAL AVE                    HOMICIDE               STREET   False     False            25.0  41.882905 -87.764950  (41.882904764, -87.764950104)
1456652    23087 2017-01-18 00:45:00            000XX W 79TH ST                    HOMICIDE               STREET   False     False            44.0  41.750934 -87.626325  (41.750934055, -87.626325022)
1456653    230

We can see that the last date is 2017-01-18, let's check at crimes in 2017 to see if there is something missing.

In [84]:
df[df['Date'].dt.year == 2017]

ID                Date                   Block Primary Type Location Description  Arrest  Domestic  Community Area   Latitude  Longitude                       Location
1444502  23059 2017-01-01 05:19:00        046XX N BROADWAY     HOMICIDE               TAVERN   False     False             3.0  41.966082 -87.657908  (41.966081547, -87.657908498)
1444503  23060 2017-01-01 06:18:00       046XX W MONROE ST     HOMICIDE               STREET   False     False            25.0  41.879291 -87.741599  (41.879290642, -87.741598513)
1446040  23061 2017-01-03 12:20:00     034XX W FULTON BLVD     HOMICIDE               STREET   False     False            27.0  41.886341 -87.712000  (41.886340707, -87.711999596)
1446041  23062 2017-01-03 23:52:00    032XX W LEXINGTON ST     HOMICIDE                ALLEY   False     False            27.0  41.871868 -87.706610  (41.871868444, -87.706610311)
1446661  23063 2017-01-04 00:41:00     034XX W FULTON BLVD     HOMICIDE               STREET   False     False            27.0  41.886341 -87.712000  (41.886340707, -87.711999596)
1446662  23064 2017-01-04 05:59:00    032XX W LEXINGTON ST     HOMICIDE                ALLEY   False     False            27.0  41.871868 -87.706610  (41.871868444, -87.706610311)
1447353  23065 2017-01-05 23:30:00         008XX E 46TH ST     HOMICIDE                 AUTO   False     False            39.0  41.811366 -87.605611  (41.811365671, -87.605610722)
1447978  23066 2017-01-06 01:43:00    086XX S MARYLAND AVE     HOMICIDE                PORCH   False     False            44.0  41.737581 -87.603634    (41.7375811, -87.603634068)
1448668  23067 2017-01-07 10:07:00     013XX W HASTINGS ST     HOMICIDE               STREET   False     False            28.0  41.864273 -87.660233  (41.864273353, -87.660232974)
1450073  23068 2017-01-09 10:07:00    045XX W LEXINGTON ST     HOMICIDE                 AUTO   False     False            26.0  41.871511 -87.738524  (41.871511269, -87.738523703)
...        ...                 ...                     ...          ...                  ...     ...       ...             ...        ...        ...                            ...
1453793  23080 2017-01-14 19:34:00         055XX W RICE ST     HOMICIDE               STREET   False     False            25.0  41.895772 -87.764223  (41.895771678, -87.764222641)
1454419  23081 2017-01-15 02:00:00         047XX W POLK ST     HOMICIDE                ALLEY   False     False            25.0  41.870473 -87.743622  (41.870472791, -87.743622093)
1454420  23082 2017-01-15 05:50:00    026XX W VAN BUREN ST     HOMICIDE               STREET   False     False            28.0  41.876028 -87.692069  (41.876028038, -87.692069498)
1455109  23084 2017-01-16 02:05:00     008XX N OAKLEY BLVD     HOMICIDE               STREET   False     False            24.0  41.896015 -87.684402  (41.896015204, -87.684402372)
1455110  23085 2017-01-16 13:35:00     004XX S CENTRAL AVE     HOMICIDE               STREET   False     False            25.0  41.874393 -87.764647   (41.874393341, -87.76464733)
1455111  23086 2017-01-16 12:11:00      048XX W HUBBARD ST     HOMICIDE                PORCH   False     False            25.0  41.888677 -87.746698  (41.888677432, -87.746697587)
1455885  23088 2017-01-17 23:42:00     001XX N CENTRAL AVE     HOMICIDE               STREET   False     False            25.0  41.882905 -87.764950  (41.882904764, -87.764950104)
1456652  23087 2017-01-18 00:45:00         000XX W 79TH ST     HOMICIDE               STREET   False     False            44.0  41.750934 -87.626325  (41.750934055, -87.626325022)
1456653  23090 2017-01-18 09:45:00  051XX S BLACKSTONE AVE     HOMICIDE               STREET   False     False            41.0  41.801809 -87.590495  (41.801809396, -87.590494612)
1456654  23091 2017-01-18 10:12:00     047XX W FLOURNOY ST     HOMICIDE               STREET   False     False            25.0  41.872293 -87.744241   (41.872292662, -87.74424056)

[30 rows x 11 columns]

Looking at the data we can see that there are only 30 rows of a total of 1.419.628 rows, since there are so few we are dropping all rows that have date of 2017.

In [87]:
df.drop(df.tail(30).index,inplace=True)
df[df['Date'].dt.year == 2017]

AttributeError: 'NoneType' object has no attribute 'drop'

In [36]:
# looks good, let's check how many classes are in each column
print("PRIMARY TYPE\n")
print(df['Primary Type'].value_counts())
print("\n\n")
print("---------------------------\n")
print("DESCRIPTION\n")
print(df['Description'].value_counts())
print("\n\n")
print("---------------------------\n")
print("LOCATION DESCRIPTION\n")
print(df['Location Description'].value_counts())
print("\n\n")
print("---------------------------\n")
print("ARREST\n")
print(df['Arrest'].value_counts())
print("\n\n")
print("---------------------------\n")
print("DOMESTIC\n")
print(df['Domestic'].value_counts())
print("\n\n")
print("---------------------------\n")
print("DISTRICT\n")
print(df['District'].value_counts())
print("\n\n")

PRIMARY TYPE

THEFT                                322433
BATTERY                              258945
CRIMINAL DAMAGE                      152817
NARCOTICS                            131184
ASSAULT                               89512
OTHER OFFENSE                         85362
BURGLARY                              81673
DECEPTIVE PRACTICE                    68356
MOTOR VEHICLE THEFT                   59858
ROBBERY                               56096
CRIMINAL TRESPASS                     36432
WEAPONS VIOLATION                     16983
PUBLIC PEACE VIOLATION                13015
OFFENSE INVOLVING CHILDREN            10591
PROSTITUTION                           7567
CRIM SEXUAL ASSAULT                    6298
INTERFERENCE WITH PUBLIC OFFICER       6136
SEX OFFENSE                            4491
HOMICIDE                               2590
GAMBLING                               2211
ARSON                                  2175
LIQUOR LAW VIOLATION                   1928
KIDNAPPING        

In [38]:
# In Primary Type we can see that we can combine Non-Criminal classes
# In Description, we can see that there are way too many classes or 340!
#   let's trim it down to 20 classes and those under are renamed to OTHER
# In location description we can see that there are 140 different classes!
#   let's trim it down to 20 classes and those under are renamed to OTHER
# Arrest, district and domestic look ok

# changing NON - CRIMINAL TO NON-CRIMINAL
df.loc[df['Primary Type'].isin(['NON - CRIMINAL', 'NON-CRIMINAL (SUBJECT SPECIFIED)']), df.columns=='Primary Type'] = 'NON-CRIMINAL'

#changing Description and Location
locToDo = list(df['Location Description'].value_counts()[20:].index)
desToDo = list(df['Description'].value_counts()[20:].index)
df.loc[df['Location Description'].isin(locToDo), df.columns=='Location Description'] = 'OTHER'
df.loc[df['Description'].isin(desToDo), df.columns=='Description'] = 'OTHER'

print("DESCRIPTION\n")
print(df['Description'].value_counts())
print("\n\n")
print("---------------------------\n")
print("LOCATION DESCRIPTION\n")
print(df['Location Description'].value_counts())
print("\n\n")

DESCRIPTION

OTHER                           356279
SIMPLE                          147850
$500 AND UNDER                  133305
DOMESTIC BATTERY SIMPLE         128285
TO VEHICLE                       74574
OVER $500                        72962
TO PROPERTY                      70343
POSS: CANNABIS 30GMS OR LESS     60273
FORCIBLE ENTRY                   52481
AUTOMOBILE                       50625
FROM BUILDING                    47541
RETAIL THEFT                     42452
UNLAWFUL ENTRY                   24340
TELEPHONE THREAT                 23016
TO LAND                          22631
ARMED: HANDGUN                   20685
POSS: HEROIN(WHITE)              20664
AGGRAVATED: HANDGUN              20201
STRONGARM - NO WEAPON            20114
HARASSMENT BY TELEPHONE          15941
CREDIT CARD FRAUD                15069
Name: Description, dtype: int64



---------------------------

LOCATION DESCRIPTION

STREET                            325091
RESIDENCE                         223858
